docs: add a genesis oracle registration docs (#613)
medibloc/panacea-core: .gitbook/5-oracles/1-operate-oracle-nodes/3-genesis-oracle.md

```shell
panacead start
```

## Registering Oracle keys and Remote Report

The genesis oracle must create an oracle private key and public key to use for data encryption/decryption.
It also issues a remote report to prove that it is a valid oracle.

### Get trusted block information
In order to generate oracle keys and remote reports, the oracle need trusted block information first. This block information is used by the light client to validate the data retrieved from Panacea.

In fact, Oracle doesn't need this block information because it doesn't retrieve Panacea's data when generating oracle keys and remote reports. 
However, when the oracle participates in the verification operation (`oracled start`), the oracle needs to use a light client as it will retrieve data from Panacea.
Therefore, unless reliable block information is received during the process of generating an oracle key, the genesis oracle has no way to retrieve this block information.

You can get trusted block information by:
```shell
BLOCK=$(panacead q block --node <node-rpc-address>)

HEIGHT=$(echo $BLOCK | jq -r .block.header.height)
HASH=$(echo $BLOCK | jq -r .block_id.hash)

> ```

Reviewer: I think this block can be put inside the below section, `Generates oracle's private key, ...`
The flow I think is,

1. let's generate oracle key
2. but for `gen-oracle-key` CLI, trusted block infromation is required. because..
3. guide to get trusted block info
4. `gen-oracle-key --trusted-block-height...`

wdyt?

Author: Yes, I think it looks good.
I will change it as your said.

### Generates oracle's private key, public key and remote report in oracle
You can generate the necessary keys and remote report via the CLI below.
```
docker run \
    --device /dev/sgx_enclave \
    --device /dev/sgx_provision \
    -v {ANY_DIR_ON_HOST}:/oracle \
    ghcr.io/medibloc/panacea-oracle:latest \
    ego run /usr/bin/oracled gen-oracle-key \

>       --trusted_block_height $HEIGHT \

Reviewer: ```suggestion
      --trusted-block-height $HEIGHT \
```

      --trusted-block-hash $HASH
      
```

| Argument             | Requirement | Description                                                 |
|----------------------|-------------|-------------------------------------------------------------|
| trusted-block-height | true        | Trusted block height of Panacea                             |
| trusted-block-hash   | true        | Block hash corresponding to trusted block height of Panacea |

> When the Oracle key and remote report generation is completed, the file is created with the following structure.

Reviewer: ```suggestion
When the oracle key and remote report are generated successfully, they are stored as file with the below structure:
```

```
# Oracle home
.
├── config.toml
├── data
│   └── light-client.db
│       ├── 000001.log
│       ├── CURRENT
│       ├── LOCK
│       ├── LOG
│       └── MANIFEST-000000
├── oracle_priv_key.sealed
└── oracle_pub_key.json
```
- `data/light-client.db`: Repository of Light client.
- `oracle_priv_key.sealed`: Oracle private key sealed file.
- `oracle_pub_key.json`: A json file containing oracle's public key and remote report.

**oracle_pub_key.json**
```json
{
  "public_key_base64": "{oracle_public_key_base64}",
  "remote_report_base64": "{oracle_remote_report_base64}"
}
```

### Submit a parameter change proposal

> In order to register Oracle's public key and remote report with Panacea, a parameter change proposal must be submit.

Reviewer: ```suggestion
The generated oracle public key and its remote report should be set by governance, a proposal for changing module parameter of oracle module.
```

```shell
panacead tx gov submit-proposal param-change proposal.json \
  --from {key} \
  --chain-id {chain_id} \
  --fees 1000000umed \
  -y
```
**proposal.json**
```json
{
  "title": "{title}",
  "description": "{description}",
  "changes": [
    {
      "subspace": "oracle",
      "key": "OraclePublicKey",
      "value": "{oracle_pub_key_base64}"
    },
    {
      "subspace": "oracle",
      "key": "OraclePubKeyRemoteReport",
      "value": "{oracle_remote_report_base64}"
    }
  ],
  "deposit": "100000000000umed"
}
```

After submitting the proposal, vote on the proposal and wait for it to pass.
```shell
panacead tx gov vote {proposal_id} yes \
  --from {key} \
  --chain-id {chain_id} \
  --fees 1000000umed \
  -y
```

If the proposal passes, you can check the changes with the following CLI.
```shell
panacead q oracle params
```
**Output**
```shell
params:
  oracle_pub_key_remote_report: "{oracle_remote_report_base64"
  oracle_public_key: "{oracle_public_key_base64}"
  unique_id: "{unique_id}"
```

> When all these processes are completed, Panacea and Oracle can operate DEP normally.

Reviewer: ```suggestion
When all these processes are completed, genesis oracle can operate normally.
```

# Genesis Oracle

- Status: Draft
- Created: 2023-01-11
- Modified: 2023-01-11
- Authors
    - Gyuguen Jang <gyuguen.jang@medibloc.org>
    - Youngjoon Lee <yjlee@medibloc.org>
    - Hansol Lee <hansol@medibloc.org>
    - Myongsik Gong <myongsik_gong@medibloc.org>
    - Inchul Song <icsong@medibloc.org>

## Synopsis
This document describes how to register a genesis oracle with Panacea.

In order for DEP to operate properly, one or more trusted oracles must be registered in Panacea.
In general, the genesis oracle will already be registered on the testnet or mainnet.
This guide is useful for testing on the localnet.

## Genesis Oracle Registration

> To register a genesis oracle, you must first complete the [oracle-inialzation](./2-oracle-intialization.md) step.

Author: Thanks. I fixed it!

```json
{
  "public_key_base64": "{oracle_public_key_base64}",
  "remote_report_base64": "{oracle_remote_report_base64}"
}
```

### Submit a parameter change proposal
The generated oracle public key and its remote report should be set by governance, a proposal for changing module parameter of oracle module.

```shell
panacead tx gov submit-proposal param-change proposal.json \
  --from {key} \
  --chain-id {chain_id} \
  --fees 1000000umed \
  -y
```
**proposal.json**
```json
{
  "title": "{title}",
  "description": "{description}",
  "changes": [
    {
      "subspace": "oracle",
      "key": "OraclePublicKey",
      "value": "{oracle_pub_key_base64}"
    },
    {
      "subspace": "oracle",
      "key": "OraclePubKeyRemoteReport",
      "value": "{oracle_remote_report_base64}"
    }
  ],
  "deposit": "100000000000umed"
}
```

After submitting the proposal, vote on the proposal and wait for it to pass.
```shell
panacead tx gov vote {proposal_id} yes \
  --from {key} \
  --chain-id {chain_id} \
  --fees 1000000umed \
  -y
```

If the proposal passes, you can check the changes with the following CLI.
```shell
panacead q oracle params
```
**Output**
```shell
params:

>   oracle_pub_key_remote_report: "{oracle_remote_report_base64"

Author: Thanks!
